crypto/threads_pthread.c

```diff
@@ -598,8 +598,16 @@ int ossl_rcu_call(CRYPTO_RCU_LOCK *lock, rcu_cb_fn cb, void *data)
      * list are visible to us prior to reading, and publish the new value
      * immediately
      */
+# ifdef __ATOMIC_ACQ_REL
     new->next = ATOMIC_EXCHANGE_N(prcu_cb_item, &lock->cb_items, new,
                                   __ATOMIC_ACQ_REL);
+# else
+    pthread_mutex_lock(&lock->write_lock);
+    new->next = lock->cb_items;
+    lock->cb_items = new;
+    pthread_mutex_unlock(&lock->write_lock);
+# endif
+
```

Comment:
This does not match to ossl_synchronize_rcu: https://github.com/openssl/openssl/blob/5949918f9afa85d72535676f114346ed541e0b1e/crypto/threads_pthread.c#L549-L552

How did this ever work?
I think this should simply use __ATOMIC_RELEASE here,
and __ATOMIC_ACQUIRE in ossl_synchronize_rcu.

Reply:
ahh, I overlooked the assignment to new->next that is a problem.

Reply:
so I think you better remove the ATOMIC_EXCHANGE completely,
and please do that also for windows.

Reply:
Note: I think the write_lock is (or should be) already held here, so this will almost certainly
dead-lock unless the mutex is "recursive"....

Reply:
> Note: I think the write_lock is (or should be) already held here, so this will almost certainly dead-lock unless the mutex is "recursive"....

Ah, right! We always call ossl_rcu_call() under the write lock. So there is no point in using atomics or any locking inside.

Reply:
> Ah, right! We always call ossl_rcu_call() under the write lock. So there is no point in using atomics or any locking inside.

Yeah, everywhere except here: https://github.com/openssl/openssl/blob/c71c65b9222135a767c39a24cb254ed792d1c942/test/lhash_test.c#L287
and there: https://github.com/openssl/openssl/blob/c71c65b9222135a767c39a24cb254ed792d1c942/test/lhash_test.c#L461
and who knows where else, however all those places are bugs, and have to be fixed as well...